pymaster/field.py

```diff
@@ -658,3 +659,240 @@ def get_ell_sampling(self):
+        self.positions = np.array(positions, dtype=np.float64)
+        self.weights = np.array(weights, dtype=np.float64)
+        if np.shape(self.positions) != (2, len(self.weights)):
+            raise ValueError("Weights must be 2D array of shape"
+                             " (2, len(weights)).")
+        if not (np.logical_and(self.positions[0] > 0.,
+                               self.positions[0] < np.pi)).all():
+            raise ValueError("First dimension of positions must be colatitude"
+                             " in radians, between 0 and pi.")
+        if not (np.logical_and(self.positions[1] > 0.,
+                               self.positions[1] < 2.*np.pi)).all():
+            raise ValueError("Second dimension of positions must be longitude"
+                             " in radians, between 0 and 2*pi.")
+
+        # 1. Compute mask alms and beam
+        # Sanity checks
+        if lmax_mask <= 0:
+            lmax_mask = hp.Alm.getlmax(np.shape(self.alm_mask)[-1])
+        self.ainfo_mask = ut.NmtAlmInfo(lmax_mask)
+        # Mask alms
+        self.alm_mask = ut._catalog2alm_ducc0(self.weights, self.positions,
+                                              spin=0, lmax=lmax_mask)[0]
+        # Beam
+        if isinstance(beam, (list, tuple, np.ndarray)):
+            if len(beam) <= lmax:
+                raise ValueError("Input beam must have at least %d elements "
+                                 "given the input map resolution" % (lmax+1))
+            beam_use = beam
+        else:
+            if beam is None:
+                beam_use = np.ones(lmax+1)
+            else:
+                raise ValueError("Input beam can only be an array or None\n")
+        self.beam = beam_use
+
+        # 2. Compute field alms
+        # If only positions and weights, just return here
+        if field is None:
+            if lmax <= 0:
+                raise ValueError("If field is None, lmax needs to be "
+                                 "provided.")
+            if spin is None:
+                raise ValueError("If field is None, spin needs to be "
+                                 "provided.")
+            self.ainfo = ut.NmtAlmInfo(lmax)
+            self.spin = spin
+            return
+        self.ainfo = ut.NmtAlmInfo(lmax)
+        self.field = np.array(field, dtype=np.float64)
+        # Sanity checks
+        if spin is None:
+            spin = 0 if field.ndim == 0 else 2
+        if spin == 2 and np.shape(self.field) != (2, len(self.weights)):
+            raise ValueError("Field has wrong shape.")
+        if spin == 0 and (self.field.ndim != 1
+                          or len(self.field) != len(self.weights)):
+            raise ValueError("Field has wrong shape.")
+        self.spin = spin
+        self.nalms = 2 if spin else 1
+        if not field_is_weighted:
+            self.field *= self.weights
+        if lmax <= 0:
+            lmax = hp.Alm.getlmax(np.shape(self.alm)[-1])
```

Comment:
As  before, `self.alm` doesn't exist yet!

Reply:
So I think we **always** need to require `lmax` for catalogs (and maybe `lmax_mask`, although this could default to `lmax` if the user leaves it as `None`).